nav2_mppi_controller/src/trajectory_visualizer.cpp

```diff
@@ -127,4 +140,33 @@ void TrajectoryVisualizer::visualize(const nav_msgs::msg::Path & plan)
   }
 }
 
+void TrajectoryVisualizer::populate_optimal_path(
+  const xt::xtensor<float, 2> & optimal_traj,
+  const builtin_interfaces::msg::Time & cmd_stamp)
+{
+  auto & size = optimal_traj.shape()[0];
+  if (size == 0) {
+    return;
+  }
+
+  optimal_path_->header.stamp = cmd_stamp;
+  optimal_path_->header.frame_id = frame_id_;
+
+  for (size_t i = 0; i < size; i++) {
+    // create new pose for the path
+    geometry_msgs::msg::PoseStamped pose_stamped;
+    pose_stamped.header.frame_id = frame_id_;
+
+    // position & orientation
+    pose_stamped.pose = utils::createPose(optimal_traj(i, 0), optimal_traj(i, 1), 0.0);
```

Comment:
This is already done in `add_marker` - would it be good to move the population of the path into that lambda? That would reduce the code duplication and looping over the data structure twice

Reply:
Indeed. `populate_optimal_path` does not exist anymore, everything is directly done in `add` and the lambda